docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,221 @@
+  - [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+Enable opt in automated health checking and remediation of unhealthy nodes backed by machines.
+
+## Motivation
+
+- Reduce administrative overhead to run a cluster.
+
+- Ability to respond to failures of machines and keep the cluster nodes healthy.
+
+### Goals
+
+- Enable automated remediation for pools of machines/nodes
+
+- Allow to define different unhealthy criterias based on node conditions for different pools.
+
+- Providing administrator-defined short-circuiting of automated remediation when multiple nodes are unhealthy at the same time.
+
+### Non-Goals/Future Work
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+The machine health checker (MHC) does a best effort to keep nodes healthy in the cluster.
+
+It provides a short-circuit mechanism and limits remediation when `maxUnhealthy` threshold is reached for a targeted pool. This is similar to what the node life cycle controller does for reducing the eviction rate as nodes goes unhealthy in a given zone. E.g a large number of nodes in a single zone are down because it's most likely a networking issue.
+
+The machine health checker is an integration point between node problem detection tooling expresed as node conditions and remediation to achieve a node auto repairing feature.
+
+
+### Unhealthy criteria:
+A machine/node target is unhealthy when:
+
+- The node meets the unhealthy node conditions criteria defined.
+- The Machine has no nodeRef.
+- The Machine has nodeRef but node is not found.
+- The Machine is in phase "Failed".
+
+If any of those criterias are met for longer than given timeouts, remediation is triggered.
+For the node conditions the time outs are defined by the admin. For the other cases opinionated values can be assumed.
+For a machine with no nodeRef an opinionated value could be assumed e.g 10 min.
+For a node notFound or a failed machine, the machine is considerable unrecoverable, remediation can be triggered right away.
+
+### Remediation:
+- The machine is requested for deletion.
+- The controller owning that machine, e.g machineSet reconciles towards the expected number of replicas and start the process to bring up a new machine/node tuple.
+- The machine controller drains the node.
+- The machine controller provider implementation deletes the cloud instance.
+- The machine controller deletes the machine resource.
+
+### User Stories
+
+#### Story 1
+
+- As a user I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated transparently in the case of failures.
+
+
+#### Story 2
+
+- As a cluster admin I want to define unhealthy criterias as a list of node conditions for targeted pools of machines running particular workloads.
```

Comment:
```suggestion
- As a cluster admin I want to define unhealthy criteria as a list of node conditions for targeted pools of machines running particular workloads.
```

Reply:
Please remove the use of pools since it could cause confusion with the `MachinePool` proposal

Reply:
Is this a Management Cluster admin or a Workload Cluster admin?